src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/Endpoints/SemanticTokens.cs

```diff
@@ -49,13 +50,48 @@ internal partial class RazorCustomMessageTarget
             return new ProvideSemanticTokensResponse(tokens: null, hostDocumentSyncVersion: csharpDoc.HostDocumentSyncVersion ?? -1);
         }
 
-        using var _ = ListPool<Task<ReinvocationResponse<VSSemanticTokensResponse>?>>.GetPooledObject(out var requestTasks);
-
         semanticTokensParams.TextDocument.Uri = csharpDoc.Uri;
         var textBuffer = csharpDoc.Snapshot.TextBuffer;
 
         cancellationToken.ThrowIfCancellationRequested();
+        ReinvocationResponse<VSSemanticTokensResponse?>? result = null;
+        if (semanticTokensParams.UsePreciseRanges)
+        {
+            var languageServerName = RazorLSPConstants.RazorCSharpLanguageServerName;
+            var lspMethodName = RazorLSPConstants.RoslynSemanticTokenRangesEndpointName;
+            var newParams = new SemanticTokensRangesParams()
+            {
+                TextDocument = semanticTokensParams.TextDocument,
+                Ranges = semanticTokensParams.Ranges
+            };
 
+            using (var disposable = _telemetryReporter.TrackLspRequest(lspMethodName, languageServerName, semanticTokensParams.CorrelationId))
+            {
+                result = await _requestInvoker.ReinvokeRequestOnServerAsync<SemanticTokensRangesParams, VSSemanticTokensResponse?>(
+                    textBuffer,
+                    lspMethodName,
+                    languageServerName,
+                    SupportsPreciseRanges,
+                    newParams,
+                    cancellationToken).ConfigureAwait(false);
+            }
+
+            // Only null if the server doesn't support the new endpoint
+            if (result != null)
+            {
+                if (result?.Response?.Data is null)
+                {
+                    _logger?.LogDebug("Made one semantic token request to Roslyn for {count} ranges but only got null result back, due to sync issues", semanticTokensParams.Ranges.Length);
+                    // Weren't able to re-invoke C# semantic tokens but we have to indicate it's due to out of sync by providing the old version
+                    return new ProvideSemanticTokensResponse(tokens: null, hostDocumentSyncVersion: csharpDoc.HostDocumentSyncVersion ?? -1);
+                }
+
+                _logger?.LogDebug("Made one semantic token requests to Roslyn for {count} ranges", semanticTokensParams.Ranges.Length);
+                return new ProvideSemanticTokensResponse(new int[][] { result?.Response?.Data! }, semanticTokensParams.RequiredHostDocumentVersion);
```

Comment:
Can this change to just take a single dimensional array of ints? I was expecting all of the stitching logic to be removed too.

Reply:
Yes the stitch code will be removed once the endpoint is available.

Reply:
+1 from me to remove the jagged array in the next PR. :smile:

Reply:
Yeah, will have to be another PR if you want this into 17.8 at all. Unfortunately it means the new code here is untested, save manual testing, but not much we can do about it unless we want to wait a few months.

Reply:
I wouldn't think the new code is not tested, why do you think that way?

Reply:
The integration tests won't run the new code until the new Roslyn bits are available on a CI image (ie, 17.8 P3 at least) and the unit tests won't run it unless you bump the version of Roslyn we reference to one that includes your changes, but doing that will likely break all of the integration tests until the CI image has that version anyway.

Reply:
Ah yea, I forgot we can't bump Roslyn because we'll break our own integration tests...

what about the rolling builds? Could definitely work on those once Roslyn inserts

Reply:
We could also try just bumping the Roslyn version used in the unit test project, and leave the rest.